File: backend/app/schemas/recent_activity.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class RecentActivityItem(BaseModel):
    stable_key: str = Field(max_length=220)
    timestamp: datetime
    actor_user_id: int | None
    actor_display: str = Field(max_length=80)
    action: str = Field(max_length=40)
    entity_type: str = Field(max_length=64)
    entity_id: int
    summary: str = Field(max_length=200)
    deep_link: str | None = Field(default=None, max_length=500)
    client_id: int | None = None
    client_name: str | None = Field(default=None, max_length=255)


class RecentActivityPage(BaseModel):
    items: list[RecentActivityItem]
    skip: int
    limit: int
    has_more: bool

File: frontend/lib/features/dashboard/domain/recent_activity.dart
class RecentActivityItem {
  const RecentActivityItem({
    required this.stableKey,
    required this.timestamp,
    required this.actorDisplay,
    required this.action,
    required this.entityType,
    required this.entityId,
    required this.summary,
    this.actorUserId,
    this.deepLink,
    this.clientId,
    this.clientName,
  });

  factory RecentActivityItem.fromJson(Map<String, dynamic> json) =>
      RecentActivityItem(
        stableKey: json['stable_key'] as String,
        timestamp: DateTime.parse(json['timestamp'] as String).toLocal(),
        actorUserId: json['actor_user_id'] as int?,
        actorDisplay: json['actor_display'] as String? ?? 'System',
        action: json['action'] as String,
        entityType: json['entity_type'] as String,
        entityId: json['entity_id'] as int,
        summary: json['summary'] as String,
        deepLink: json['deep_link'] as String?,
        clientId: json['client_id'] as int?,
        clientName: json['client_name'] as String?,
      );

  final String stableKey;
  final DateTime timestamp;
  final int? actorUserId;
  final String actorDisplay;
  final String action;
  final String entityType;
  final int entityId;
  final String summary;
  final String? deepLink;
  final int? clientId;
  final String? clientName;
}

class RecentActivityPageData {
  const RecentActivityPageData({
    required this.items,
    required this.skip,
    required this.limit,
    required this.hasMore,
  });

  factory RecentActivityPageData.fromJson(Map<String, dynamic> json) =>
      RecentActivityPageData(
        items: (json['items'] as List<dynamic>? ?? const <dynamic>[])
            .whereType<Map<String, dynamic>>()
            .map(RecentActivityItem.fromJson)
            .toList(growable: false),
        skip: json['skip'] as int? ?? 0,
        limit: json['limit'] as int? ?? 8,
        hasMore: json['has_more'] as bool? ?? false,
      );

  final List<RecentActivityItem> items;
  final int skip;
  final int limit;
  final bool hasMore;
}
